source/opt/decoration_manager.cpp

```diff
@@ -39,24 +126,84 @@ std::vector<const ir::Instruction*> DecorationManager::GetDecorationsFor(
       ->InternalGetDecorationsFor<const ir::Instruction*>(id, include_linkage);
 }
 
-// TODO(pierremoreau): The code will return true for { deco1, deco1 }, { deco1,
-//                     deco2 } when it should return false.
 bool DecorationManager::HaveTheSameDecorations(uint32_t id1,
                                                uint32_t id2) const {
-  const auto decorationsFor1 = GetDecorationsFor(id1, false);
-  const auto decorationsFor2 = GetDecorationsFor(id2, false);
-  if (decorationsFor1.size() != decorationsFor2.size()) return false;
-
-  for (const ir::Instruction* inst1 : decorationsFor1) {
-    bool didFindAMatch = false;
-    for (const ir::Instruction* inst2 : decorationsFor2) {
-      if (AreDecorationsTheSame(inst1, inst2, true)) {
-        didFindAMatch = true;
-        break;
+  using InstructionList = std::vector<const ir::Instruction*>;
+
+  const InstructionList decorationsFor1 = GetDecorationsFor(id1, false);
+  const InstructionList decorationsFor2 = GetDecorationsFor(id2, false);
+
+  // This function will sort and remove all duplicates for each of the three
+  // given lists.
+  const auto generateUniqueLists = [](const InstructionList& decorationList,
+                                      InstructionList* decorateList,
```

Comment:
If you just need to remove duplicates and compare the content, then you could use
std::unordered_set<std::u32string>
This way you won't need most of the custom code.

Reply:
Hum, and sorting to avoid a quadratic behaviour wouldn’t be necessary as we would rely on the hash map of the unordered set? I can do the change. Why an unordered set of `std::u32string` though (I didn’t even know that type existed), rather than a std::string?